### PR TITLE
[KYUUBI #2686] Fix lock bug if engine initialization timeout

### LINKED ISSUE #2686: [Bug] Fix lock bug if engine initialization timeout
### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### Search before asking

- [X] I have searched in the [issues](https://github.com/apache/incubator-kyuubi/issues?q=is%3Aissue) and found no similar issues.

### Describe the bug

We should throw exception if timeout during acquiring lock.
Let's say we have three clients with same request lock to two kyuubi server instances.
        
 client A  --->  kyuubi X  -- first acquired  \
 client B  --->  kyuubi X  -- second acquired --  zookeeper
 client C  --->  kyuubi Y  -- third acquired  /
        
The first client A acqiured the lock then B and C are blocked until A release the lock,
with the A created engine state:
  - SUCCESS
    B acquired the lock then get engine ref and release the lock.
    C acquired the lock then get engine ref and release the lock.
  - FAILED or TIMEOUT
    B acquired the lock then try to create engine again.
    C should be timeout and throw exception back to client. This fast fail
    to avoid client too long to waiting in concurrent.

Also different engine type should use its own lock.

### Affects Version(s)

master/1.5

### Kyuubi Server Log Output

_No response_

### Kyuubi Engine Log Output

_No response_

### Kyuubi Server Configurations

_No response_

### Kyuubi Engine Configurations

_No response_

### Additional context

_No response_

### Are you willing to submit PR?

- [x] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by ulysses-you**
### _Why are the changes needed?_

closes #2686

### _How was this patch tested?_
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [ ] [Run test](https://kyuubi.apache.org/docs/latest/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

File: kyuubi-server/src/test/scala/org/apache/kyuubi/engine/EngineRefSuite.scala [hunk 1]
```diff
@@ -247,4 +248,81 @@ class EngineRefSuite extends KyuubiFunSuite {
...
+        }
+      }
+    })
+
+    eventually(timeout(10.seconds), interval(200.milliseconds)) {
+      assert(times.forall(_ > start))
+      // ENGINE_INIT_TIMEOUT is 3000ms
+      assert(times.max - times.min < 2500)
+    }
+    executor.shutdown()
```
Comment (turboFei) [inline kyuubi-server/src/test/scala/org/apache/kyuubi/engine/EngineRefSuite.scala:292]: should we shutdown the launched engines?

Comment (turboFei) [reply kyuubi-server/src/test/scala/org/apache/kyuubi/engine/EngineRefSuite.scala:292]: how about adding `afterEach` method to close all the launched engines?

Comment (turboFei) [reply kyuubi-server/src/test/scala/org/apache/kyuubi/engine/EngineRefSuite.scala:292]: seems afterAll will do that, please ignore this.

Comment (ulysses-you) [reply kyuubi-server/src/test/scala/org/apache/kyuubi/engine/EngineRefSuite.scala:292]: engines have already shutdown due to the short `ENGINE_INIT_TIMEOUT`

Comment (turboFei) [reply kyuubi-server/src/test/scala/org/apache/kyuubi/engine/EngineRefSuite.scala:292]: ok

**Comment by ulysses-you** [comment]
unbelievable my local works well with the test code... but the ga failed sometime

**Comment by ulysses-you** [comment]
Seems the test is stable, cc @yaooqinn @turboFei @pan3793 @cfmcgrady

File: kyuubi-server/src/test/scala/org/apache/kyuubi/engine/EngineRefSuite.scala [hunk 2]
```diff
@@ -205,6 +208,8 @@ class EngineRefSuite extends KyuubiFunSuite {
     conf.set(ENGINE_POOL_SIZE, 3)
     val engine6 = new EngineRef(conf, user, id, null)
     assert(engine6.subdomain.startsWith(s"$enginePoolName-"))
+    conf.unset(ENGINE_POOL_NAME)
+    conf.unset(ENGINE_POOL_SIZE)
```
Comment (turboFei) [inline kyuubi-server/src/test/scala/org/apache/kyuubi/engine/EngineRefSuite.scala:212]: seems not needed because you have added it into `beforeEach`

**Review by turboFei — APPROVED** [review]

**Comment by yaooqinn** [comment]
what is the current behavior？

**Comment by ulysses-you** [comment]
> what is the current behavior？

if three(more than two) request coming, if the first request initializes engine timeout, we will get two engines from the last two request.

**Comment by yaooqinn** [comment]
if A hangs with a big timeout， considering B/C are sequentially retries of A， so they are going to make server fe threads be exhausted？

**Comment by ulysses-you** [comment]
> > > what is the current behavior？
> > 
> > if three(more than two) request coming, if the first request initializes engine timeout, we will get two engines from the last two request.
> 
> hmm？ so the timeout here means time to acquire the lock，not time to hold the lock？

yeah.. that's right

**Comment by yaooqinn** [comment]
so，shall we throw exception if timeout？

File: kyuubi-server/src/test/scala/org/apache/kyuubi/engine/EngineRefSuite.scala (hunk 2 above)
Comment (ulysses-you) [reply kyuubi-server/src/test/scala/org/apache/kyuubi/engine/EngineRefSuite.scala]: yeah, let me remove this first to see any issue in test

File: kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/zookeeper/ZookeeperDiscoveryClient.scala [hunk 3]
```diff
@@ -143,9 +143,34 @@ class ZookeeperDiscoveryClient(conf: KyuubiConf) extends DiscoveryClient {
...
+        //   - SUCCESS
+        //     B acquired the lock then get engine ref and release the lock.
+        //     C acquired the lock then get engine ref and release the lock.
+        //   - FAILED or TIMEOUT
+        //     B acquired the lock then try to create engine again if not timeout.
+        //     C should be timeout and throw exception back to client. This fast fail
+        //     to avoid client too long to waiting in concurrent.
+
+        // Return false means we are timeout
+        val acquired = lock.acquire(timeout, unit)
```
Comment (yaooqinn) [inline kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/zookeeper/ZookeeperDiscoveryClient.scala:166]: use TimeUnit.MILLISECONDS directly, the timeout is always millis

**Review by hddong — APPROVED** [review]

File: kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/zookeeper/ZookeeperDiscoveryClient.scala (hunk 3 above)
Comment (ulysses-you) [reply kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/zookeeper/ZookeeperDiscoveryClient.scala]: removed

File: kyuubi-server/src/test/scala/org/apache/kyuubi/engine/EngineRefSuite.scala [hunk 4]
```diff
@@ -247,4 +256,84 @@ class EngineRefSuite extends KyuubiFunSuite {
...
+
+    eventually(timeout(20.seconds), interval(200.milliseconds)) {
+      assert(times.forall(_ > start))
+      // ENGINE_INIT_TIMEOUT is 3000ms
+      assert(times.max - times.min > 2800)
+    }
+
+    // we should only submit two engines, the last request should timeout and fail
+    assert(MetricsSystem.counterValue(ENGINE_TOTAL).get - beforeEngines == 2)
+    executor.shutdown()
```
Comment (yaooqinn) [inline kyuubi-server/src/test/scala/org/apache/kyuubi/engine/EngineRefSuite.scala:337]: shall we put it in the finally block?

File: kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/zookeeper/ZookeeperDiscoveryClient.scala (hunk 3 above)
Comment (hddong) [reply kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/zookeeper/ZookeeperDiscoveryClient.scala]: Can we keep the unit here? ETCD lease timeout is seconds, and cannot change unit.

Comment (yaooqinn) [reply kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/zookeeper/ZookeeperDiscoveryClient.scala]: simply timeout / 1000 for etcd?

Comment (hddong) [reply kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/zookeeper/ZookeeperDiscoveryClient.scala]: It's OK to

File: kyuubi-server/src/test/scala/org/apache/kyuubi/engine/EngineRefSuite.scala (hunk 4 above)
Comment (ulysses-you) [reply kyuubi-server/src/test/scala/org/apache/kyuubi/engine/EngineRefSuite.scala]: yeah, added it

**Review by yaooqinn — APPROVED** [review]

**Comment by ulysses-you** [comment]
thanks for review ! merging to master/branch-1.5

**Comment by ulysses-you** [comment]
seems a lot of conflicts..  I will do backport manually

**Pull request closed without merging**